crates/gen/src/handle.rs

```diff
@@ -2,11 +2,32 @@ use super::*;
 
 pub fn gen_handle() -> TokenStream {
     quote! {
+        #[derive(::std::clone::Clone, ::std::marker::Copy, ::std::default::Default, ::std::fmt::Debug, ::std::cmp::PartialEq, ::std::cmp::Eq)]
+        #[repr(transparent)]
+        pub struct HANDLE(pub isize);
         impl HANDLE {
+            pub const NULL: Self = Self(0);
             pub const INVALID: Self = Self(-1);
             pub fn is_invalid(&self) -> bool {
                 self.0 == -1
             }
         }
+        unsafe impl ::windows::Handle for HANDLE {
+            fn is_null(&self) -> bool {
+                *self == unsafe { ::std::mem::zeroed() }
+            }
+
+            fn ok(self) -> ::windows::Result<Self> {
+                if self != Self::NULL && self != Self::INVALID {
+                    Ok(self)
+                } else {
+                    Err(::windows::HRESULT::from_thread().into())
```

Comment:
I think this will return spurious HRESULTs from APIs that don't `SetLastError`, right?

Reply:
Yes, the caller is responsible for using it as appropriate. That's why it's a method call and not simply woven into the projection.

Reply:
The part that bugs me a bit is that checking for null would be common operation but its result on error would probably be _wrong_ most of the time.

Wonder if we should just make this return a bool and the caller can explicitly retrieve the hresult as needed elsewhere?

Just a nit.

Reply:
You can use is_null for that. I expect that to be more common. The ok method is consistent with methods of the same name across types like HRESULT and BOOL that all return Result.